src/ILCompiler.TypeSystem/tests/GenericTypeAndMethodTests.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Internal.TypeSystem;
+using Xunit;
+
+namespace TypeSystemTests
+{
+    public class GenericMethodTests
+    {
+        private TestTypeSystemContext _context;
+        private ModuleDesc _testModule;
+
+        public GenericMethodTests()
+        {
+            _context = new TestTypeSystemContext(TargetArchitecture.Unknown);
+            var systemModule = _context.CreateModuleForSimpleName("CoreTestAssembly");
+            _context.SetSystemModule(systemModule);
+
+            _testModule = systemModule;
+        }
+
+        /// <summary>
+        /// Testing proper instantiation of methods involving generic parameters in their signature.
+        /// </summary>
+        [Fact]
+        [ActiveIssue(-1)]
+        public void TestMethodInstantiation()
+        {
+            MetadataType t = _testModule.GetType("GenericTypes", "GenericClass`1");
+            MethodDesc fooMethod = t.GetMethods().First(m => m.Name == "Foo");
+            MethodDesc instantiatedMethod =
```

Comment:
I think this is why I created the test in the first place, it is surprising to make the difference between the two. What you are describing is the notion of a class vs a type. A type should always be instantiated whereas a class doesn't need to be.

In any case, if this is the indented behavior, can you confirm the following and I'll update the test accordingly next week?
* In the above test, `t` is a generic type definition where `IsTypeDefinition` returns true, `HasInstantiation` returns false and `ContainsGenericVariables` returns true. One has to use `MakeInstantiatedType` to get a generic type Instantiation.
* Calling InstantiateSignature on a generic type definition should do nothing and returns the same method.